File: app/src/main/java/com/ilifesmart/ui/SimpleGroup.java
package com.ilifesmart.ui;

import android.content.Context;
import android.graphics.Canvas;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;

public class SimpleGroup extends ViewGroup {

	public static final String TAG = "SimpleGroup";

	public SimpleGroup(Context context) {
		super(context);
	}

	public SimpleGroup(Context context, AttributeSet attrs) {
		super(context, attrs);
	}

	public SimpleGroup(Context context, AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);
	}

	@Override
	protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
		int widthSize = MeasureSpec.getSize(widthMeasureSpec);
		int widthMode = MeasureSpec.getMode(widthMeasureSpec);

		int heightSize = MeasureSpec.getSize(heightMeasureSpec);
		int heightMode = MeasureSpec.getMode(heightMeasureSpec);

		int height = 0;
		int maxWidth = 0;

		Log.d(TAG, "onMeasure: count " + getChildCount());
		for (int index = 0; index < getChildCount(); index++) {
			View child = getChildAt(index);
			measureChildWithMargins(child, widthMeasureSpec, 0, heightMeasureSpec, 0);
			MarginLayoutParams lp = (MarginLayoutParams) child.getLayoutParams();

			height += child.getMeasuredHeight() + lp.bottomMargin + lp.topMargin;
			Log.d(TAG, "onMeasure: height " + height);
			maxWidth = Math.max(maxWidth, child.getMeasuredWidth());
		}

		int nWidth = (widthMode == MeasureSpec.EXACTLY) ? widthSize : maxWidth;
		int nHeight= (heightMode == MeasureSpec.EXACTLY)? heightSize: height;
		setMeasuredDimension(nWidth, nHeight);
	}

	@Override
	protected void onLayout(boolean changed, int l, int t, int r, int b) {
		int top = 0;
		int count = getChildCount();

		Log.d(TAG, "onLayout: count " + count);
		for (int index = 0; index < count; index++) {
			View child = getChildAt(index);
			MarginLayoutParams lp = (MarginLayoutParams) child.getLayoutParams();

			int height = child.getMeasuredHeight() + lp.topMargin + lp.bottomMargin;
			int width = child.getMeasuredWidth();
			child.layout(0, top, width, top + height); // 先onSizeChanged后onLayout

			top += height;
		}
	}

	@Override
	protected LayoutParams generateLayoutParams(LayoutParams p) {
		return new MarginLayoutParams(p);
	}

	@Override
	public LayoutParams generateLayoutParams(AttributeSet attrs) {
		return new MarginLayoutParams(getContext(), attrs);
	}

	@Override
	protected LayoutParams generateDefaultLayoutParams() {
		return new MarginLayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT);
	}

	@Override
	protected void onDraw(Canvas canvas) {
		super.onDraw(canvas);
	}

	@Override
	protected void dispatchDraw(Canvas canvas) {
		super.dispatchDraw(canvas);
	}


}

File: app/src/main/java/com/ilifesmart/ui/SimpleGroup.kt
package com.ilifesmart.ui

import android.animation.Animator
import android.animation.ObjectAnimator
import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.os.Handler
import android.os.Looper
import android.util.AttributeSet
import android.util.Log
import android.view.View
import android.view.ViewGroup

import com.ilifesmart.utils.DensityUtils
import com.ilifesmart.utils.Utils

class SimpleGroup : ViewGroup {
    private var animator: ValueAnimator? = null

    private var once = true

    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val widthSize = View.MeasureSpec.getSize(widthMeasureSpec)
        val widthMode = View.MeasureSpec.getMode(widthMeasureSpec)

        val heightSize = View.MeasureSpec.getSize(heightMeasureSpec)
        val heightMode = View.MeasureSpec.getMode(heightMeasureSpec)

        var maxHeight = 0
        var width = 0

        for (index in 0 until childCount) {
            val child = getChildAt(index)
            measureChildWithMargins(child, widthMeasureSpec, 0, heightMeasureSpec, 0)
            val lp = child.layoutParams as ViewGroup.MarginLayoutParams

            maxHeight = Math.max(child.measuredHeight + lp.bottomMargin + lp.topMargin, maxHeight)
            width += child.measuredWidth + lp.leftMargin + lp.rightMargin
        }

        val nWidth = if (widthMode == View.MeasureSpec.EXACTLY) widthSize else width
        val nHeight = if (heightMode == View.MeasureSpec.EXACTLY) heightSize else maxHeight
        setMeasuredDimension(nWidth, nHeight)
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        var left = 0
        val count = childCount

        Log.d(TAG, "onLayout: count $count")
        for (index in 0 until count) {
            val child = getChildAt(index)
            val lp = child.layoutParams as ViewGroup.MarginLayoutParams

            val height = child.measuredHeight + lp.topMargin + lp.bottomMargin
            val width = child.measuredWidth
            child.layout(left, 0, left + width, height) // 先onSizeChanged后onLayout

            left += width
        }
    }

    override fun generateLayoutParams(p: ViewGroup.LayoutParams): ViewGroup.LayoutParams {
        return ViewGroup.MarginLayoutParams(p)
    }

    override fun generateLayoutParams(attrs: AttributeSet): ViewGroup.LayoutParams {
        return ViewGroup.MarginLayoutParams(context, attrs)
    }

    override fun generateDefaultLayoutParams(): ViewGroup.LayoutParams {
        return ViewGroup.MarginLayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()

        animator!!.cancel()
        animator = null
    }

    fun autoPlay() {
        if (childCount == 0 || animator != null && animator!!.isRunning) {
            return
        }

        val arr = IntArray(2)
        DensityUtils.getWindowSize(context, arr)
        val windowWidth = arr[0]
        animator = ValueAnimator.ofInt(0, windowWidth)

        animator!!.duration = 4000
        animator!!.startDelay = 4000
        animator!!.addUpdateListener { animation ->
            val value = animation.animatedValue as Int
            for (i in 0 until childCount) {
                getChildAt(i).translationX = (-value).toFloat()
            }
        }

        animator!!.addListener(object : Animator.AnimatorListener {
            override fun onAnimationStart(animation: Animator) {
                if (!once) {
                    val lastIndex = 0 //(loopCount++) % getChildCount();
                    val v = getChildAt(lastIndex)
                    removeViewAt(lastIndex)
                    addView(v)
                    requestLayout()
                }
                once = false
            }

            override fun onAnimationEnd(animation: Animator) {
                animation.start()
            }

            override fun onAnimationCancel(animation: Animator) {}

            override fun onAnimationRepeat(animation: Animator) {
                Log.d(TAG, "onAnimationRepeat: ")
            }
        })
        animator!!.start()
    }

    companion object {

        val TAG = "SimpleGroup"
    }

}
